### PR TITLE
Next-gen member-ordering rule

### LINKED ISSUE #1205: Member-ordering issue with constuctors
### Bug Report

- __TSLint version__: `3.9.0`
- __TypeScript version__: `1.8.9`
- __Running TSLint via__: Gulp

#### TypeScript code being linted

```ts
export class SkeletonDirective {
	/*@ngInject*/
	constructor(
	) {
		// todo: constructor injection and dependency initialization here
	}

	static factory(): IDirectiveFactory {
		/*@ngInject*/
		const directive = (
		) => {
			return new SkeletonDirective();
		};
		return directive;
	}
};
```

with `tslint.json`:

```json
		"member-ordering": [
			true,
			"public-before-private",
			"static-before-instance",
			"variables-before-functions"
		],
```

#### Actual behavior

Declaration of public static member function not allowed to appear after declaration of public instance member function

#### Expected behavior

No warning as it is a constuctor

### PR DESCRIPTION
**Pull request by JKillian**
Adds an optional new syntax for more explicit control of the `member-ordering` rule (see #1208).

This should resolve #1205 and #1186 by giving a user more power to define their own orderings. This keeps backwards-compatibility with the old rule options (which will be removed in TSLint 4.0). This also allows for the possibility of further fields, like an `alphabetical: true` option.

### CODE REVIEW

File: src/rules/memberOrderingRule.ts [hunk 1]
```diff
@@ -49,36 +88,97 @@ function toString(modifiers: IModifiers): string {
...
-export class MemberOrderingWalker extends Lint.RuleWalker {
-    private previousMember: IModifiers;
+/* start new code */
+enum AccessLevel {
+    PRIVATE,
+    PROTECTED,
+    PUBLIC,
+}
+
+enum Bound {
```
Comment (JKillian) [inline src/rules/memberOrderingRule.ts]: Is this the best name? Any ideas on better names?

File: src/rules/memberOrderingRule.ts [hunk 2]
```diff
@@ -17,16 +17,55 @@
...
 
+/* start old options */
 const OPTION_VARIABLES_BEFORE_FUNCTIONS = "variables-before-functions";
 const OPTION_STATIC_BEFORE_INSTANCE = "static-before-instance";
 const OPTION_PUBLIC_BEFORE_PRIVATE = "public-before-private";
+/* end old options */
+
+/* start new options */
+const OPTION_ORDER = "order";
+const PRESET_ORDERS: { [preset: string]: string[] } = {
```
Comment (JKillian) [inline src/rules/memberOrderingRule.ts]: @lowkay @thasner @alan-agius4 (and others) would you use one of these presets? If not, what preset would you like?

Comment (lowkay) [inline src/rules/memberOrderingRule.ts]: @jkillian Yupp, fields-first would be my preference.

Comment (alan-agius4) [inline src/rules/memberOrderingRule.ts]: Most likely yes and I'd go for fields-first.

Comment (JKillian) [inline src/rules/memberOrderingRule.ts]: @marines also

File: src/rules/memberOrderingRule.ts (hunk 1 above)
Comment (marines) [inline src/rules/memberOrderingRule.ts]: How about?
```java
enum Membership {
    CLASS,
    INSTANCE
}
```

File: src/rules/memberOrderingRule.ts [hunk 3]
```diff
@@ -17,16 +17,70 @@
...
+        "private-static-method",
+        "public-instance-field",
+        "protected-instance-field",
+        "private-instance-field",
+        "constructor",
+        "public-instance-method",
+        "protected-instance-method",
+        "private-instance-method",
+    ],
+    "instance-sandwich": [
```
Comment (JKillian) [inline src/rules/memberOrderingRule.ts]: added this for you @thasner

File: src/rules/memberOrderingRule.ts (hunk 2 above)
Comment (marines) [inline src/rules/memberOrderingRule.ts]: Looks like the instance-sandwich will suit me the best but as long as I can order the rules by myself I'm totally fine with that.

File: src/rules/memberOrderingRule.ts (hunk 1 above)
Comment (JKillian) [inline src/rules/memberOrderingRule.ts]: I like that better, thanks!

File: src/rules/memberOrderingRule.ts (hunk 3 above)
Comment (thasner) [inline src/rules/memberOrderingRule.ts]: 👍

**Comment by adidahiya** [comment]
code lgtm, but needs tests for the other presets

**Comment by JKillian** [comment]
> code lgtm, but needs tests for the other presets
Think so? I was thinking that it wasn't really necessary to have tests for every preset

File: test/rules/member-ordering/order/fields-first/test.ts.lint [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    constructor() {}
+    public static g() {}
+    protected static h() {}
+    private static i = () => {};
+    j = () => {};
+    protected k() {};
+    private l = () => {};
+}
+    
+Interface IGood {
```
Comment (adidahiya) [inline test/rules/member-ordering/order/fields-first/test.ts.lint]: typo, `interface`

this is likely not being linted

**Comment by adidahiya** [comment]
hm, ok, you're right that we don't need to test all the presets, but I'd like some more tests that check that access level modifiers are enforced

**Comment by adidahiya** [comment]
👍

**Pull request merged**